core/src/rpc/aggregator.rs

```diff
@@ -111,44 +112,8 @@ impl ClementineAggregator for Aggregator {
         tracing::debug!("Parsed deposit params");
 
         // generate nonces from all verifiers
-        let mut nonce_streams = try_join_all(self.verifier_clients.iter().map(|client| {
-            // Clone each client to avoid mutable borrow.
-            // https://github.com/hyperium/tonic/issues/33#issuecomment-538150828
-            let mut client = client.clone();
-
-            async move {
-                let response_stream = client.nonce_gen(Request::new(Empty {})).await?;
-
-                Ok::<_, Status>(response_stream.into_inner())
-            }
-        }))
-        .await?;
-
-        tracing::debug!("Generated nonce streams");
-
-        // Get the first responses from each stream
-        let first_responses = try_join_all(nonce_streams.iter_mut().map(|s| async {
-            let nonce = s.message().await?;
-            let pub_nonce_response = nonce
-                .ok_or(Status::internal("No nonce received"))?
-                .response
-                .ok_or(Status::internal("No nonce received"))?;
-            // this response is an enum, so we need to match on it
-            match pub_nonce_response {
-                nonce_gen_response::Response::FirstResponse(nonce_gen_first_response) => {
-                    Ok::<_, Status>(nonce_gen_first_response)
-                }
-                _ => panic!("Expected FirstResponse"),
-            }
-        }))
-        .await
-        .map_err(|e| {
-            Status::internal(format!("Failed to get first nonce gen responses {:?}", e))
-        })?;
-
-        tracing::debug!("Received first responses: {:?}", first_responses);
-
-        let num_required_nonces = first_responses[0].num_nonces as usize; // TODO: This should be the same for all verifiers
+        let (first_responses, mut nonce_streams) =
+            self.create_nonce_streams(NUM_REQUIRED_SIGS as u32).await?;
```

Comment:
As suggested above, it is better that RPC message passing things should stay at the RPC crate.